### PR TITLE
Allow acme boolean flag to enable lego behavior with default issuer

### PR DESCRIPTION
**Pull request by fabn**
Signed-off-by: Fabio Napoleoni <f.napoleoni@gmail.com>

## Description

Implement a `--acme` flag for cert-manager install subcommand that automates helm values 

```
--set=ingressShim.defaultIssuerName=letsencrypt-prod \
--set=ingressShim.defaultIssuerKind=ClusterIssuer
```

and create a global `ClusterIssuer` that is used to generate letsencrypt certificates when an ingress is annotated with `kubernetes.io/tls-acme: "true"`

Let me know if you're interested and if I should change something to make this merged.

## Motivation and Context

As explained in #271 it would be useful to have an app specific flag that implements [`kubernetes.io/tls-acme: "true"`](https://cert-manager.io/docs/usage/ingress/#optional-configuration) configuration.

I think is in the motivation of this project to have some shorthand to automate per cluster tasks that are tedious and error prone like the one above.

Also the implementation is completely optin.

## How Has This Been Tested?

I built the binary locally and tested on my own cluster, I'm not so skilled with go (actually my first lines in go) thus I'm still not able to implement an automated test. If you give me some hint on how to do that I'll be glad to try.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [x] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

**Comment by Waterdrips** [comment]
Hey, we like to discuss changes like this on issues, I know we had one to discuss the missing `--set` flag from cert-manager app. I think you also proposed adding the functionality in this PR but it didnt really get addressed.

I'm not sure of the implementation of this feature, we have some apps creating LE issuers (openfaas-ingress and docker-registry-ingress). They can install a staging or production issuer with different ingressClass annotations (traefik for example) and they have templated the YAML used to create the issuer, this lets us do the above modifications as well as not require using yaml from the internet (which may change between cli versions). 

Additionally the `--acme` flag you have added is true by default, which is either a breaking change (if someone has an issuer called that already) or will start to install cluster-scoped resources into people's clusters without them knowing.

**Comment by fabn** [comment]
Good points, I'll try to answer:

* true by default was obviously a mistake, my proposed change should be optin. I'll change it
* for the yaml part I completely agree that that the proposed approach is not so clean and my first idea was to embed some kind of  yaml template in the command, also because that will allow further customization (e.g. to register a letsencrypt email address in issuer) but as I told you my go skills are pretty limited. If you point me to the right direction I'll change it to use a template instead of downloading the issuer template from git repo.

**Comment by fabn** [comment]
I've update PR with suggested changes, let me know if they aree good.

**Comment by fabn** [comment]
@Waterdrips feedback?

**Comment by Waterdrips** [comment]
Looks good, can we add a --staging flag like we have for the other apps that create issuers? 

Good for testing and not using your rate limit up.

**Comment by fabn** [comment]
@Waterdrips done. I have a test failure but is unrelated to my changes. I think is because of openfaas new releasees.

**Comment by fabn** [comment]
@Waterdrips up

**Review by Waterdrips — CHANGES_REQUESTED** [review]
2 things,
1. can you update the tests to look for the new faas-cli verison.
2. the issuer secret needs to be named after the correct issuer (stg/prod)

Also, the commits need squashing down into 1 commit which should have a good commit message https://chris.beams.io/posts/git-commit/

**Comment by fabn** [comment]
Done, please check it out.

File: cmd/apps/certmanager_app.go [hunk 1]
```diff
@@ -95,6 +111,13 @@ func MakeInstallCertManager() *cobra.Command {
 			overrides["installCRDs"] = "true"
 		}
 
+		// Allow usage of lego behavior, see https://cert-manager.io/docs/usage/ingress/#optional-configuration
+		useAcme, _ := certManager.Flags().GetBool("acme")
+		if useAcme {
+			overrides["ingressShim.defaultIssuerName"] = "letsencrypt-production-issuer"
```
Comment (Waterdrips) [inline cmd/apps/certmanager_app.go:117]: I think this needs the same staging logix as the issuer as here, https://github.com/alexellis/arkade/pull/272/files#diff-90097b757fed057b84f0204d467ed1708d81bdf17fc9e0188fd9c22fcc04b577R219-R222

So the sataging issuer name is used if --staging is set?

**Review by Waterdrips — CHANGES_REQUESTED** [review]
one small change for the overrides map, need to use staging if --staging is set

File: cmd/apps/certmanager_app.go (hunk 1 above)
Comment (fabn) [reply cmd/apps/certmanager_app.go:117]: You're right, that's needed as well.

I'll change it

Comment (fabn) [reply cmd/apps/certmanager_app.go:117]: Done.

**Comment by fabn** [comment]
Up

**Comment by Shikachuu** [comment]
@fabn are you still intrested, to move the PR forward?

**Comment by fabn** [comment]
I used this fork on more than one cluster and it works as is. What should I do to get it merged?

File: cmd/apps/certmanager_app.go [hunk 2]
```diff
@@ -129,7 +173,94 @@ func MakeInstallCertManager() *cobra.Command {
...
+	if templateErr != nil {
+		log.Print("Unable to install the application. Could not build the templated yaml file for the resources")
+		return templateErr
+	}
+
+	tempFile, tempFileErr := writeTempFile(yamlBytes, "temp_registry_ingress.yaml")
+	if tempFileErr != nil {
+		log.Print("Unable to save generated yaml file into the temporary directory")
+		return tempFileErr
+	}
```
Comment (Shikachuu) [inline cmd/apps/certmanager_app.go:197]: Imo it would be more reasonable to `fmt.Errorf` the log line, like in line 206 or every other application.

**Review by Shikachuu — CHANGES_REQUESTED** [review]
Everything except the log lines looks good to me. (Also thats just a code style issue tho'.)

File: cmd/apps/certmanager_app.go (hunk 2 above)
Comment (fabn) [reply cmd/apps/certmanager_app.go]: Done

**Comment by Shikachuu** [comment]
@fabn can you rebase your commits and sign-off your commit please?

**Comment by fabn** [comment]
> @fabn can you rebase your commits and sign-off your commit please?

Done

**Comment by fabn** [comment]
Also fixed some compile error when rebased.

**Comment by Shikachuu** [comment]
LGTM

**Review by Shikachuu — APPROVED** [review]

**Pull request closed without merging**